rootsanalyzer/rootsanalyzer_test.go

```diff
@@ -0,0 +1,206 @@
+			rootSetIDs: []storage.RootSetID{rootSet1ID, rootSet2ID},
+			rootSetCerts: map[storage.RootSetID][]*x509.Certificate{
+				rootSet1ID: {root1, root2},
+				rootSet2ID: {root1},
+			},
+			wantReport: &itestonly.Report{
+				Summary:  "Root certificates changed",
+				Category: "roots",
+				BaseURL:  "https://ct.googleapis.com/testtube/",
+				FullURL:  "https://ct.googleapis.com/testtube/ct/v1/get-roots",
+				Details: `The root certificates accepted by testtube (https://ct.googleapis.com/testtube/) have changed.
+
+Certificates removed (1):
+CN=A-Trust-Qual-02,OU=A-Trust-Qual-02,O=A-Trust Ges. f. Sicherheitssysteme im elektr. Datenverkehr GmbH,C=AT
+`,
+			},
+		},
+		{
+			desc:       "2 certs removed",
+			rootSetIDs: []storage.RootSetID{rootSet1ID, rootSet2ID},
+			rootSetCerts: map[storage.RootSetID][]*x509.Certificate{
+				rootSet1ID: {root1, root2},
+				rootSet2ID: {},
+			},
+			wantReport: &itestonly.Report{
+				Summary:  "Root certificates changed",
+				Category: "roots",
+				BaseURL:  "https://ct.googleapis.com/testtube/",
+				FullURL:  "https://ct.googleapis.com/testtube/ct/v1/get-roots",
+				Details: `The root certificates accepted by testtube (https://ct.googleapis.com/testtube/) have changed.
+
+Certificates removed (2):
+CN=A-Trust-Qual-02,OU=A-Trust-Qual-02,O=A-Trust Ges. f. Sicherheitssysteme im elektr. Datenverkehr GmbH,C=AT
+CN=GLOBALTRUST,OU=GLOBALTRUST Certification Service,O=ARGE DATEN - Austrian Society for Data Protection,L=Vienna,ST=Austria,C=AT
+`,
+			},
+		},
+		{
+			desc:       "1 cert added, 1 cert removed",
+			rootSetIDs: []storage.RootSetID{rootSet1ID, rootSet2ID},
+			rootSetCerts: map[storage.RootSetID][]*x509.Certificate{
+				rootSet1ID: {root1},
+				rootSet2ID: {root2},
+			},
+			wantReport: &itestonly.Report{
+				Summary:  "Root certificates changed",
+				Category: "roots",
+				BaseURL:  "https://ct.googleapis.com/testtube/",
+				FullURL:  "https://ct.googleapis.com/testtube/ct/v1/get-roots",
+				Details: `The root certificates accepted by testtube (https://ct.googleapis.com/testtube/) have changed.
+
+Certificates added (1):
+CN=A-Trust-Qual-02,OU=A-Trust-Qual-02,O=A-Trust Ges. f. Sicherheitssysteme im elektr. Datenverkehr GmbH,C=AT
+
+Certificates removed (1):
+CN=GLOBALTRUST,OU=GLOBALTRUST Certification Service,O=ARGE DATEN - Austrian Society for Data Protection,L=Vienna,ST=Austria,C=AT
+`,
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			test := test
+			t.Parallel()
+
+			fakeStorage := &stestonly.FakeRootsReader{
+				RootSetChan:  make(chan storage.RootSetID),
+				RootSetCerts: test.rootSetCerts,
+			}
+			fakeReporter := &itestonly.FakeReporter{
+				Reports: make(chan itestonly.Report),
+			}
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+
+			go Run(ctx, fakeStorage, fakeReporter, l)
+
+			for _, id := range test.rootSetIDs {
+				fakeStorage.RootSetChan <- id
+			}
```

Comment:
Am I right in thinking that there's a limit in the number of ids you could have in test.rootSetIDs before the channels start blocking things up a bit?  Looking at it, it might actually be 2, but that works because your tests only ever need 2, as they currently just check between 2 root set IDs?

Mostly to check I understand channels etc :P

Reply:
The answer is actually 0 - it's an unbuffered channel, so sending to the channel blocks until a reader comes along. I start the analyzer just before this though, so it should be reading and unblock the channel. If it isn't, the test will timeout in this loop. There's no limit on how many IDs you can have in `test.rootSetIDs`, since the analyzer should read them all. However, I have just realised that it'll block when it uses the incident reporter, since that also uses an unbuffered channel, so that would always have to be the last thing that happened. To avoid this, I've made them both buffered channels of appropriate size.